File: Coding-Patterns/Coding-Patterns-Python/02_trees/lowest_common_ancestor.py
# LeetCode 236 - Lowest Common Ancestor of a Binary Tree
# Difficulty: Medium
# Pattern: DFS post-order
#
# Problem: Given a binary tree and two nodes p and q,
# find their lowest common ancestor (LCA).
#
# Approach: Post-order DFS. If current node is p or q, return it.
# If both left and right subtrees return non-null, current node is LCA.
# Otherwise propagate the non-null result upward.
#
# Time: O(n)  Space: O(h)

from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def lowest_common_ancestor(root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
    if not root or root == p or root == q:
        return root

    left = lowest_common_ancestor(root.left, p, q)
    right = lowest_common_ancestor(root.right, p, q)

    if left and right:
        return root   # p and q are in different subtrees
    return left or right  # both in same subtree


# ---- Tests ----
if __name__ == "__main__":
    #        3
    #       / \
    #      5   1
    #     / \ / \
    #    6  2 0  8
    #      / \
    #     7   4
    n6, n7, n4, n0, n8 = TreeNode(6), TreeNode(7), TreeNode(4), TreeNode(0), TreeNode(8)
    n2 = TreeNode(2, n7, n4)
    n5 = TreeNode(5, n6, n2)
    n1 = TreeNode(1, n0, n8)
    root = TreeNode(3, n5, n1)

    assert lowest_common_ancestor(root, n5, n1).val == 3
    assert lowest_common_ancestor(root, n5, n4).val == 5
    print("All tests passed.")

File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/02_trees/LowestCommonAncestor.cs
// LeetCode 236 - Lowest Common Ancestor of a Binary Tree
// Difficulty: Medium
// Pattern: DFS post-order
//
// Approach: Post-order DFS. Return node if it matches p or q.
// If both left and right are non-null, current node is LCA.
//
// Time: O(n)  Space: O(h)

namespace CodingPatterns.Trees;

public class LowestCommonAncestor
{
    public TreeNode LCA(TreeNode root, TreeNode p, TreeNode q)
    {
        if (root == null || root == p || root == q) return root;
        var left = LCA(root.Left, p, q);
        var right = LCA(root.Right, p, q);
        if (left != null && right != null) return root;
        return left ?? right;
    }

    // ---- Tests ----
    public static void Run()
    {
        var sol = new LowestCommonAncestor();

        //        3
        //       / \
        //      5   1
        //     / \
        //    6   2
        var six = new TreeNode(6);
        var two = new TreeNode(2);
        var five = new TreeNode(5, six, two);
        var one = new TreeNode(1);
        var root = new TreeNode(3, five, one);

        Console.WriteLine(sol.LCA(root, five, one).Val);   // 3
        Console.WriteLine(sol.LCA(root, six, two).Val);    // 5
        Console.WriteLine(sol.LCA(root, five, six).Val);   // 5 (a node is its own ancestor)
    }
}
